### PR TITLE
OPENEUROPA-1374: Demonstrate the corporate editorial workflow

### PR DESCRIPTION
**Pull request by nagyad**
### Description

Created demo content type with extra configuration in order to provide the full editorial functionality out of the box, just by enabling the oe_editorial_workflow_demo module. I created BEHAT tests for the corporate editorial worfklow.

I had to add "symfony/browser-kit": "~4.2.1", because it was removed from composer_lowest build and behat could not be executed without.

### CODE REVIEW

File: tests/Behat/FeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Defines step definitions that are generally useful in this project.
+ */
+class FeatureContext extends RawDrupalContext {
+
+  /**
+   * Checks that the given select field has the options listed in the table.
+   *
+   * @Then I should have the following options for :select select:
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: for *the* :select select :)

Also can we maybe put a little doc on how this step would be used? Kinda like some of the methods inside `Drupal\DrupalExtension\Context\DrupalContext`?

File: tests/Behat/FeatureContext.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    // Check that the specified field is a <select> field.
+    $this->assertElementType($field, 'select');
+
+    // Retrieve the options table from the test scenario and flatten it.
+    $expected_options = $options->getRows();
+    array_walk($expected_options, function (&$value) {
+      $value = reset($value);
+    });
+
+    // Retrieve the actual options that are shown in the page.
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: *on* the *select field* (not page).

File: tests/Behat/FeatureContext.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    // Convert into a flat list of option text strings.
+    array_walk($actual_options, function (&$value) {
+      $value = $value->getText();
+    });
+
+    // Check that all expected options are present.
+    foreach ($expected_options as $expected_option) {
+      if (!in_array($expected_option, $actual_options)) {
+        throw new ExpectationException("Option '$expected_option' is missing from select list '$select'.", $this->getSession());
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: Maybe use `sprintf()`?

File: tests/Behat/FeatureContext.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+   *
+   * @param \NodeElement $element
+   *   The element to check.
+   * @param string $type
+   *   The expected type.
+   *
+   * @throws \ExpectationException
+   *   Thrown when the given element is not of the expected type.
+   */
+  public function assertElementType(NodeElement $element, $type): void {
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: This should be a protected method. And I would move it to the bottom. We don't have a rule but I think it's neat when all the public methods are at the top (steps) and helper protected methods at the bottom.

File: tests/Behat/FeatureContext.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    if ($element->getTagName() !== $type) {
+      throw new ExpectationException("The element is not a '$type'' field.", $this->getSession());
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: sprintf() ? :)

File: tests/Behat/FeatureContext.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+   *
+   * @param string $state
+   *   The expected workflow state.
+   *
+   * @throws \Exception
+   *   Thrown when the current workflow state field is not shown on the page.
+   *
+   * @Then the current workflow state should be :state
+   */
+  public function assertCurrentWorkflowState($state): void {
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: Type hint.

File: tests/Behat/FeatureContext.php (hunk 4 above)
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: Type hint

File: tests/Behat/FeatureContext.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ * Defines step definitions that are generally useful in this project.
+ */
+class FeatureContext extends RawDrupalContext {
+
+  /**
+   * Checks that the given select field has the options listed in the table.
+   *
+   * @Then I should have the following options for :select select:
+   */
+  public function assertSelectOptions($select, TableNode $options): void {
```
Comment (upchuk) [inline tests/Behat/FeatureContext.php]: Type hint.

File: tests/features/editorial_workflow.feature [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    Given I am logged in as a user with the "Author" role
+    When I visit "node/add/oe_workflow_demo"
+    Then I should have the following options for "Save as" select:
+      | Draft        |
+      | Needs Review |
+    When I fill in "Title" with "Workflow demo"
+    And I press "Save"
+    And I visit "admin/content"
+    # The content is created and it is not published.
+    Then I should see text matching "Workflow demo"
```
Comment (upchuk) [inline tests/features/editorial_workflow.feature]: Even if in this case you only have one node, this should be a step that check that on that specific table row the status is "not published". You can use the `DrupalContext::assertTextInTableRow`

File: modules/oe_editorial_workflow_demo/oe_editorial_workflow_demo.install [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        $drupal_role->grantPermission($permission);
+        $drupal_role->save();
+      }
+    }
+  }
+}
+
+/**
+ * Implements hook_uninstall().
+ */
```
[the remark this answers is not in the document]
Comment (nagyad) [reply modules/oe_editorial_workflow_demo/oe_editorial_workflow_demo.install]: The permissions coming with the content type, these are generic permissions, not custom. By uninstalling the module the content type will be removed with their permissions as well, right?

File: modules/oe_editorial_corporate_workflow/README.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+## Setup workflow for content types
+
+In order to be able to use the functionality you need to do the following:
+
+* Create a content type
+* Assign the content type specific permissions to the editorial roles (eg.: Content type X: Create new content)
+* Go to "admin/config/workflow/workflows" and edit the Corporate workflow and assign your 
+content type under "THIS WORKFLOW APPLIES TO:" to this workflow
+
+After executing these steps as a site administrator, you should have the workflow applied to your desired content type.
```
Comment (hernani) [inline modules/oe_editorial_corporate_workflow/README.md]: We should also mention that content moderation can be applied to other entity types as long as they are revisionable. That is currently the case for content types and media types. Taxonomy terms are currently not revisionable in 8.6 but might be in 8.7. RDF Entity do not support currently revisions and therefore cannot use yet any content workflow.

**Pull request merged**